app/objects/secondclass/c_link.py

```diff
@@ -226,31 +226,49 @@ async def _load_module(module_type, module_info):
-            await self._save_fact(operation, relationship.source, relationship.score)
-            await self._save_fact(operation, relationship.target, relationship.score)
+            relationship.origin = operation.id if operation else self.id
+            await self._save_fact(operation, relationship.source, relationship.score, relationship.shorthand)
+            await self._save_fact(operation, relationship.target, relationship.score, relationship.shorthand)
             if all((relationship.source.trait, relationship.edge)):
+                knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+                await knowledge_svc_handle.add_relationship(relationship)
                 self.relationships.append(relationship)
 
-    async def _save_fact(self, operation, fact, score):
-        all_facts = operation.all_facts() if operation else self.facts
-        if all([fact.trait, fact.value]) and await self._is_new_fact(fact, all_facts):
-            self.facts.append(Fact(trait=fact.trait, value=fact.value, score=score, collected_by=self.paw,
-                                   technique_id=self.ability.technique_id))
-
-    async def _is_new_fact(self, fact, facts):
-        return all(not self._fact_exists(fact, f) or self._is_new_host_fact(fact, f) for f in facts)
-
-    @staticmethod
-    def _fact_exists(new_fact, fact):
-        return new_fact.trait == fact.trait and new_fact.value == fact.value
-
-    def _is_new_host_fact(self, new_fact, fact):
-        return new_fact.trait[:5] == 'host.' and self.paw != fact.collected_by
-
-    async def _update_scores(self, operation, increment):
-        for uf in self.used:
-            all_facts = operation.all_facts() if operation else self.facts
-            for found_fact in all_facts:
-                if found_fact.unique == uf.unique:
-                    found_fact.score += increment
-                    break
+    async def _save_fact(self, operation, fact, score, relationship):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+        all_facts = await operation.all_facts() if operation else self.facts
+        source = operation.id if operation else self.id
+        rl = [relationship] if relationship else []
+        if all([fact.trait, fact.value]):
+            if not await knowledge_svc_handle.check_fact_exists(fact, all_facts):
+                f_gen = Fact(trait=fact.trait, value=fact.value, source=source, score=score, collected_by=self.paw,
+                             technique_id=self.ability.technique_id, links=[self.id], relationships=rl,
+                             origin_type=OriginType.LEARNED)
+                self.facts.append(f_gen)
+                await knowledge_svc_handle.add_fact(f_gen)
+            else:
+                existing_fact = (await knowledge_svc_handle.get_facts(criteria=dict(trait=fact.trait,
+                                                                                    value=fact.value)))[0]
```

Comment:
why not use `fact.unique` here if it is just trait and value? (even though I proposed changing that above..)
And since we are taking the first index, is that going to be the right fact, or do we want to adjust the links on the proper source fact?

Reply:
The knowledge service is supposed to de-conflict identical facts, so you should theoretically only ever get one fact back here.  I could definitely see moving to the use of fact.unique, but you'd need to generate it on the fly, and if we ever change the value of fact.unique, that will break everything (see previous reply/statement).

Reply:
ok, thats what I was thinking.  And I can see what you mean on not using fact.unique, though it seems we do have a use case for a property on facts of just trait and value, so I could see adding something specifically for that.

Reply:
Actually, looking at this more, I think it (and the line that follows in a few lines down), should also be looking for a match on the fact's source. Since that's the case, I'm going to add that, rather than make something that can be a single lookup field on the fact for just the trait and value. Eventually we can come back and replace this with whatever we migrate unique to.

Reply:
sounds good, and I was figuring that as well, as you always want to add the relationships to the comparable fact for the proper source.